Fix for [General] "MetaMask is Locked" error message: Use async functions for accounts and network.

cryptomental opened the pull request:
Fixes: #111


##### Description
Fixes MetaMask detection by using async versions of accounts and network.

##### Checklist

- [X] Linter status: 100% pass
- [X] Changes don't break existing behavior
- [X] Tests coverage hasn't decreased

##### Affected core subsystem(s)
web3 init

##### Testing
Tested manually.

##### Refers/Fixes
Fixes: #111

pelsasser commented:
This looks to be WAY better!
@eswarasai can merge once need testing is fixed.

cryptomental commented:
@perfectmak I fixed network but cannot figure out the mocks for getAccounts and getNetwork. Perhaps I am too tired today. It looks like the web3 fake provider does not provide async versions and we have to mock them. I tried ```Object.assign(mockWindow, {
      web3: {
        currentProvider: provider,
        eth: {
          accounts,
          getAccounts: jest.fn().mockImplementation(cb => { accounts })
        },
        version: {
          getNetwork: jest.fn().mockImplementation(cb => { network: '4447' })
        },
      },
    });``` but the promises do not return correctly. Would be grateful if you could have a look. Apart from that I tested many times the fix and it works without issues.

eswarasai commented:
@cryptomental -- So @perfectmak had raised a PR for adding tests for `actions`. Is [this](https://github.com/perfectmak/Dapp/blob/9c9291894eac373add3692463ef5683f80d5b20c/test/actions/deploy.test.js#L23) something you're looking for?

PS: I can merge the PR #107 if needed here. Thanks.

cryptomental commented:
@eswarasai thanks, yes I suspect this is what I need.

cryptomental commented:
If there is no progress ill have a look at this again tomorrow morning.

cryptomental commented:
@eswarasai @perfectmak @pelsasser thanks for merging Perfect's PR, I rebased my changes on top and refactored the web3 tests to use properly mocked web3 with mocked getAccounts() and getNetwork() callback functions. This PR can me merged now.

eswarasai commented on src/util/web3/getWeb3.js:
```diff
@@ -38,23 +38,23 @@ let getWeb3 = (window,
 
         showErrorMessage('MetaMask is locked! Please unlock and refresh this page', 8);
         resolve(dispatch(web3Initialized(results)));
         return;
```
`return` is redundant here. Can be removed.

cryptomental replied:
good catch!

eswarasai commented on src/util/web3/getWeb3.js:
```diff
@@ -66,14 +66,14 @@ let getWeb3 = (window,
-        }
-      }
-
+        };
```
`if` block doesn't need a semi-colon. And also instead of having empty `return`, we can move the `return network` to else block and remove the `return` statement from `if` block?

cryptomental replied:
ok

eswarasai commented on src/util/web3/getWeb3.js:
```diff
@@ -38,23 +38,23 @@ let getWeb3 = (window,
-          web3.version.network !== RINKEBY &&
-          web3.version.network !== TRUFFLE) { // ensure beta users don't spend real ether.
+      const network = web3.version.getNetwork((error, network) => {
+        if (network !== RINKEBY &&
```
Can we have all the checks in a single line here instead of line breaks for the network conditions?

cryptomental replied:
on it.
